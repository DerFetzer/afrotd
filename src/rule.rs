use eyre::eyre;
use maud::{html, PreEscaped, Render};
use roman_numerals::ToRoman;
use std::{fmt::Display, str::FromStr};

#[derive(Debug, Clone, Hash, PartialEq, Eq)]
pub struct Rule {
    pub article_nr: ArticleNr,
    pub title: String,
    pub text: String,
    pub interpretations: Vec<RuleInterpretation>,
}

impl Rule {
    #[allow(clippy::comparison_chain)]
    fn render_text(&self) -> maud::Markup {
        let mut current_indent = 0u8;
        let processed_lines = self.text.lines().map(|l| {
            let (new_indent, list_type) = if l.starts_with("\t\t\t") {
                (3, "a")
            } else if l.starts_with("\t\t") {
                (2, "1")
            } else if l.starts_with('\t') {
                (1, "a")
            } else {
                (0, "")
            };
            let penalty_index = l.find(" Strafe: ");

            let markup = html! {
                @if new_indent > current_indent {
                    (PreEscaped(format!("<ol type={list_type}>")))
                } @else if new_indent < current_indent {
                    @for _ in 0..(current_indent - new_indent){
                        (PreEscaped("</ol>"))
                    }
                }
                @if new_indent != 0 && !l.starts_with("\tAusnahmen") {
                    li { ({
                        let space_index = l.find(' ').expect("Missing space after enumeration index.");
                        if let Some(penalty_index) = penalty_index {
                            html! {
                                p { (&l[space_index..penalty_index + 1]) }
                                p { strong { (&l[penalty_index + 1..]) } }
                            }
                        } else {
                            html! {
                                ( &l[space_index..] )
                            }
                        }
                    }) }
                } @else if let Some(penalty_index) = penalty_index {
                    p { (l[..penalty_index + 1]) }
                    p { strong { (l[penalty_index + 1..]) } }
                } @else if l.starts_with("\tAusnahmen") {
                    p { strong { (l) } }
                } @else {
                    p { (l) }
                }
            };
            current_indent = new_indent;
            markup
        });
        html! {
            .content {
                @for line in processed_lines {
                    (line)
                }
            }
        }
    }

    pub fn to_description(&self) -> String {
        format!(
            "{}...",
            self.text
                .lines()
                .next()
                .unwrap()
                .chars()
                .take(50)
                .collect::<String>()
        )
    }

    pub fn to_title(&self) -> String {
        format!("{} {}", self.article_nr, self.title)
    }

    pub fn to_url(&self, base_url: &str) -> String {
        format!("{}/rule/{}", base_url, self.article_nr.to_path_parameter())
    }
}

impl Render for Rule {
    fn render(&self) -> maud::Markup {
        html! {
            article.message ."is-size-4" {
                div.message-header {
                    p { (self.article_nr) " " (self.title) }
                }
                div.message-body {
                    (self.render_text())
                    div.block {
                        a .button .is-medium
                            href=(format!("https://afsvd.de/content/files/2024/12/Football_Regelbuch_2025.pdf#{}", self.article_nr.to_pdf_destination()))
                            target="_blank" rel="noreferrer noopener" {
                            "Offizielles Regelwerk"
                        }
                    }
                    div.block {
                        @for interpretation in &self.interpretations {
                            (interpretation)
                        }
                    }
                }
            }
        }
    }
}

#[derive(Debug, Clone, Hash, PartialEq, Eq)]
pub struct RuleInterpretation {
    pub article_nr: ArticleNr,
    pub index: u8,
    pub text: String,
    pub ruling: String,
}

impl Render for RuleInterpretation {
    fn render(&self) -> maud::Markup {
        html! {
            article.message ."is-size-5" .is-info {
                div.message-header {
                    p { "A.R. " (self.article_nr) "." (self.index.to_roman()) }
                }
                div.message-body {
                    p { (self.text) }
                    p {
                        details {
                            summary { b { "Regelung" } }
                            p {
                                (self.ruling)
                            }
                        }
                    }
                }
            }
        }
    }
}

#[derive(Debug, Copy, Clone, Hash, PartialEq, Eq)]
pub struct ArticleNr(pub u8, pub u8, pub u8);

impl ArticleNr {
    pub fn from_path_paramter(path_parameter: String) -> eyre::Result<Self> {
        let article_nr_parts = path_parameter
            .splitn(3, '-')
            .map(|p| p.parse())
            .collect::<Result<Vec<u8>, _>>()
            .map_err(|_| eyre!("Parts have to be integer"))?;
        if let [chapter, section, article] = &article_nr_parts[..] {
            Ok(ArticleNr(*chapter, *section, *article))
        } else {
            Err(eyre!("Parameter has to consist of three parts"))
        }
    }

    pub fn to_pdf_destination(self) -> String {
        if self.2 != 0 {
            format!("subsection.1.{}.{}.{}", self.0, self.1, self.2)
        } else {
            format!("section.1.{}.{}", self.0, self.1)
        }
    }

    pub fn to_path_parameter(self) -> String {
        format!("{}-{}-{}", self.0, self.1, self.2)
    }
}

impl Display for ArticleNr {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}.{}.{}", self.0, self.1, self.2)
    }
}

impl FromStr for ArticleNr {
    type Err = eyre::Report;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let parts: Vec<_> = s.trim().trim_end_matches('.').split('.').collect();
        if let [chapter, section, article] = &parts[..] {
            Ok(Self(chapter.parse()?, section.parse()?, article.parse()?))
        } else {
            Err(eyre!("Invalid article number: {s}"))
        }
    }
}
