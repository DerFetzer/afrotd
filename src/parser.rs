use eyre::{eyre, Context};
use indexmap::IndexMap;
use once_cell::sync::Lazy;
use regex::{Captures, Regex};
use roman_numerals::FromRoman;
use std::{path::Path, process::Command};

use crate::rule::{ArticleNr, Rule, RuleInterpretation};

pub struct RulesParser;

impl RulesParser {
    pub fn parse(rules_path: &Path) -> eyre::Result<IndexMap<ArticleNr, Rule>> {
        let pdftotext_output = Command::new("pdftotext")
            .args([
                "-x",
                "0",
                "-y",
                "40",
                "-W",
                "1000",
                "-H",
                "540",
                rules_path
                    .to_str()
                    .ok_or(eyre!("Invalid path to rules pdf"))?,
                "-",
            ])
            .output()
            .wrap_err("Could not run pdftotext")?;
        if !pdftotext_output.status.success() {
            return Err(eyre!(
                "pdftotext did not exit successfuly: {}",
                String::from_utf8(pdftotext_output.stderr)?
            ));
        }
        let mut rules_text =
            String::from_utf8(pdftotext_output.stdout).wrap_err("Stdout is no valid utf8")?;

        rules_text = Self::preprocess_text(rules_text);

        let mut rules = Self::extract_rules(&rules_text)?;
        let interpretations = Self::extract_interpretations(rules_text)?;

        for (article_nr, article_interpretations) in interpretations {
            let rule = rules
                .get_mut(&article_nr)
                .ok_or_else(|| eyre!("Could not find rule {}", article_nr))?;
            rule.interpretations = article_interpretations;
        }

        Ok(rules)
    }

    fn preprocess_text(mut text: String) -> String {
        // Treat section 9.1. as rule
        text = text.replace(
            "Abschnitt 9.1 Persönliche Fouls\nAlle",
            "Artikel 9.1.0 Persönliche Fouls\nAlle",
        );
        text = text.replace("Regel 9\nVerhalten von Spielern und anderen", "");

        // Fixes for rules that are longer than one line
        text = text.replace("9.1.4 Targeting und Forcible Contact zum Kopf-/Halsbereich\nverteidigungsloser Spieler",
                                        "9.1.4 Targeting und Forcible Contact zum Kopf-/Halsbereich verteidigungsloser Spieler");
        text = text.replace(
            "6.1.3 Berühren, illegales Berühren und Recovern eines Free\nKicks",
            "6.1.3 Berühren, illegales Berühren und Recovern eines Free Kicks",
        );

        let re_new_page = Regex::new(r"-?\n\x0C").unwrap();
        let re_section = Regex::new(r"(?s)Abschnitt .*?Artikel").unwrap();
        let re_new_chapter = Regex::new(r"\n\x0C.*\n.*\n\n").unwrap();

        text = re_new_page.replace_all(&text, "").to_string();
        text = re_section.replace_all(&text, "\nArtikel").to_string();
        text = re_new_chapter.replace_all(&text, "").to_string();

        text
    }

    fn extract_rules(text: &str) -> eyre::Result<IndexMap<ArticleNr, Rule>> {
        let rules_start = text
            .find("Artikel 1.1.1")
            .ok_or(eyre!("Could not find 'Artikel 1.1.1' inside the pdf text"))?;
        let rules_end = text
            .find("Die Abkürzungen R, Ab, Art stehen für Regel,")
            .ok_or(eyre!(
                "Could not find 'Die Abkürzungen R, Ab, Art stehen für Regel,' inside the pdf text"
            ))?;

        let mut rules = IndexMap::new();

        let re_article_header =
            Regex::new(r"(?m)Artikel (?<article_nr>\d+\.\d+\.\d+) (?<title>.*)$").unwrap();

        let rules_part = &text[rules_start..rules_end];

        let captures: Vec<_> = re_article_header.captures_iter(rules_part).collect();

        let mut last_captures = None;

        for (article_header, next) in captures.iter().zip(captures.iter().skip(1)) {
            last_captures = Some(next);
            let rule = Self::extract_rule_from_text(
                rules_part,
                article_header,
                next.get(0).unwrap().start(),
            )?;
            rules.insert(rule.article_nr, rule);
        }
        if let Some(last_capture) = last_captures {
            let end_of_last_capture = last_capture.get(0).unwrap().end();
            let rules_end = rules_part[end_of_last_capture..]
                .find("Zusammenfassung der Strafen")
                .ok_or(eyre!("Could not find end of rules"))?
                + end_of_last_capture;
            let rule = Self::extract_rule_from_text(rules_part, last_capture, rules_end)?;
            rules.insert(rule.article_nr, rule);
        }
        dbg!(rules.len());
        Ok(rules)
    }

    fn extract_interpretations(
        text: String,
    ) -> eyre::Result<IndexMap<ArticleNr, Vec<RuleInterpretation>>> {
        let re_rule = Regex::new(r"(?sm)^Regel .*?A\.R\.").unwrap();
        let re_section = Regex::new(r"(?sm)^Abschnitt .*?A\.R\.").unwrap();
        let re_article = Regex::new(r"(?sm)^Artikel .*?A\.R\.").unwrap();

        let mut text = re_rule.replace_all(&text, "\nA.R.").to_string();
        text = re_section.replace_all(&text, "\nA.R.").to_string();
        text = re_section.replace_all(&text, "\nA.R.").to_string();
        text = re_article.replace_all(&text, "\nA.R.").to_string();

        let interpretations_start = text.find("\nA.R. 1.3.2.I ").ok_or(eyre!(
            "Could not find '\\nA.R. 1.3.2.I ' inside the pdf text"
        ))?;
        let interpretations_end = text
            .find("Teil III")
            .ok_or(eyre!("Could not find 'Teil III' inside the pdf text"))?;

        let mut interpretations: IndexMap<ArticleNr, Vec<RuleInterpretation>> = IndexMap::new();

        let re_interpretation = Regex::new(r"(?sm)^A\.R\. (?<ar_nr>\d+\.\d+\.\d+.)(?<index>[IVX]+) (?<situation>.*?)Regelung:(?<ruling>.*?)\nA\.R\. ").unwrap();

        let mut interpretations_text = text[interpretations_start..interpretations_end].to_string();
        interpretations_text.push_str("\nA.R. "); // Add interpretation header to match last one
        let mut current_position = 0;

        while let Some(captures) =
            re_interpretation.captures_at(&interpretations_text, current_position)
        {
            let article_nr = captures["ar_nr"].parse()?;
            interpretations
                .entry(article_nr)
                .or_default()
                .push(RuleInterpretation {
                    article_nr,
                    index: u8::from_roman(&captures["index"])
                        .ok_or_else(|| eyre!("Invalid Roman number."))?,
                    text: captures["situation"].trim().to_string(),
                    ruling: captures["ruling"].trim().to_string(),
                });
            current_position = captures.get(0).unwrap().end() - 5;
        }

        Ok(interpretations)
    }

    fn extract_rule_from_text(
        text: &str,
        article_header: &Captures,
        next_start: usize,
    ) -> eyre::Result<Rule> {
        static RE_NEWLINE_ALPHA_LISTING: Lazy<Regex> =
            Lazy::new(|| Regex::new(r"\n([a-z]\) )").unwrap());
        static RE_NEWLINE_NUM_LISTING: Lazy<Regex> =
            Lazy::new(|| Regex::new(r"\n([0-9]+\. )").unwrap());
        static RE_NEWLINE_INNER_ALPHA_LISTING: Lazy<Regex> =
            Lazy::new(|| Regex::new(r"\n([a-z]\. )").unwrap());
        static RE_NEWLINE_WITHOUT_TAB: Lazy<Regex> =
            Lazy::new(|| Regex::new(r"\n([^\t])").unwrap());

        static RE_EXCEPTIONS: Lazy<Regex> = Lazy::new(|| Regex::new(r"\n(Ausnahmen:\n)").unwrap());
        static RE_TRAILING_NUM: Lazy<Regex> = Lazy::new(|| Regex::new(r" \d+$").unwrap());

        let article_nr: ArticleNr = article_header["article_nr"].parse()?;
        let title = article_header["title"].parse()?;

        let article_text = &text[article_header.get(0).unwrap().start()..next_start];

        let text_start = article_text
            .find('\n')
            .ok_or(eyre!("No newline after header"))?;
        let mut text = article_text[text_start..].to_string();

        // Special handling for Clipping and Blocking in the back
        if (title == "Clipping" && article_nr.0 == 9) || title == "Blocken in den Rücken" {
            text = RE_NEWLINE_ALPHA_LISTING
                .replace_all(&text, "\n\t\t\t$1")
                .to_string();
        } else {
            text = RE_NEWLINE_ALPHA_LISTING
                .replace_all(&text, "\n\t$1")
                .to_string();
        }
        text = RE_NEWLINE_NUM_LISTING
            .replace_all(&text, "\n\t\t$1")
            .to_string();
        text = RE_NEWLINE_INNER_ALPHA_LISTING
            .replace_all(&text, "\n\t\t\t$1")
            .to_string();

        text = RE_EXCEPTIONS.replace_all(&text, "\n\t$1").to_string();

        text = RE_NEWLINE_WITHOUT_TAB
            .replace_all(&text, " $1")
            .trim()
            .to_string();

        text = RE_TRAILING_NUM.replace_all(&text, "").to_string();

        // Special replacements for Targeting rule
        text = text.replace("Anmerkung 1 Targeting", "\nAnmerkung 1\nTargeting");
        text = text.replace(
            "Anmerkung 2 Verteidigungslose",
            "\nAnmerkung 2\nVerteidigungslose",
        );

        if text.starts_with("a)") {
            text = format!("\t{}", text);
        }

        Ok(Rule {
            article_nr,
            title,
            text,
            interpretations: vec![],
        })
    }
}
