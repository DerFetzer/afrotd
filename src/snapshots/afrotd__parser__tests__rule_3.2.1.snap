---
source: src/parser.rs
expression: rule
snapshot_kind: text
---
article_nr:
  - 3
  - 2
  - 1
title: Länge der Perioden und der Pausen
text: "Die reine Spielzeit in einem Spiel beträgt 48 Minuten, geteilt in vier Perioden zu je 12 Minuten (Ausnahme: Jugendspiele siehe BSO, 2. Damenbundesliga siehe ligeninterne Regelung). Zwischen der ersten und zweiten Periode (erste Halbzeit) und der dritten und vierten Periode (zweite Halbzeit) soll eine Pause von einer Minute liegen. (Ausnahme: Diese 1-Minuten-Pausen können für Radio- und TV-Timeouts verlängert werden.)\n\ta) Keine Periode darf beendet werden, bevor der Ball dead ist und der Referee die Periode für beendet erklärt hat [S14].\n\tb) Die Pause zwischen den Halbzeiten eines regulären Saisonspieles soll 15 Minuten dauern. Die Pause kann in gemeinsamen Einverständnis der Managements beider Teams verkürzt werden. Unmittelbar nach dem Ende der zweiten Spielperiode soll der Referee anzeigen, dass die Halbzeitpause beginnt [S2].\n\tc) Am Ende der ersten Halbzeit, nachdem die Teams das Feld verlassen haben und der Referee (a) den letzten Down mit der SchiedsrichterCrew abgestimmt und (b) festgestellt hat, dass es keine Einwände eines Head Coaches gibt, kann er die Halbzeit für beendet erklären. Nachdem der Referee die erste Halbzeit für beendet erklärt hat, werden keine zusätzlichen Einwände von dem vorherigen Down zugelassen.\n\td) Das Heimteammanagement bestimmt den Zeitpunkt der Halbzeitaktivitäten. Das Spielfeld steht den Spielteilnehmern spätestens drei Minuten vor Anpfiff der zweiten Halbzeit zur Verfügung. Wenn ein Spielteilnehmer während der Halbzeitpause den Platz betritt, muss ein Verantwortlicher dieser Mannschaft auf dem Platz ebenfalls anwesend sein. Wenn Kicker oder andere Spielteilnehmer den Platz betreten, bevor das Spielfeld verfügbar ist, müssen ihre Aktivitäten auf die eigene Teamzone beschränkt werden. Alle Spielteilnehmer müssen die geplanten Halbzeitaktivitäten respektieren und alle Kicks müssen in die Kick-Netze erfolgen, bis das Spielfeld wieder verfügbar ist."
interpretations: []
